src/modules/0.0.11/aws_iam/index.ts

```diff
@@ -1,10 +1,99 @@
-import { Role as AWSRole } from '@aws-sdk/client-iam'
+import {
+  CreateRoleCommandInput,
+  GetRoleCommandOutput,
+  IAM,
+  ListAttachedRolePoliciesCommandOutput,
+  paginateListRoles,
+  Role as AWSRole
+} from '@aws-sdk/client-iam'
 
 import { Role } from './entity'
-import { AWS, } from '../../../services/gateways/aws_2'
+import { AWS, crudBuilder, mapLin, paginateBuilder, } from '../../../services/aws_macros'
 import { Context, Crud2, Mapper2, Module2, } from '../../interfaces'
 import * as metadata from './module.json'
-import logger from '../../../services/logger'
+
+const getRoleAttachedPoliciesArns = crudBuilder<IAM>(
+  'listAttachedRolePolicies',
+  (roleName: string) => ({ RoleName: roleName, }),
+  (res: ListAttachedRolePoliciesCommandOutput) => res.AttachedPolicies?.length ? res.AttachedPolicies.map(p => p.PolicyArn ?? '') : undefined,
+);
+
+// TODO: How to macro-ify this passing the attachedPolicyArns as argument for the return formatter?
+async function createNewRole(client: IAM, input: CreateRoleCommandInput, attachedPolicyArns: string[]): Promise<string> {
+  const role = await client.createRole(input);
+  for (const arn of attachedPolicyArns) {
+    await client.attachRolePolicy({PolicyArn: arn, RoleName: input.RoleName});
+  }
+  return role.Role?.Arn ?? '';
+}
+
+const createInstanceProfile = crudBuilder<IAM>(
+  'createInstanceProfile',
+  (roleName: string) => ({ InstanceProfileName: roleName, }),
+);
+
+const attachRoleToInstanceProfile = crudBuilder<IAM>(
+  'addRoleToInstanceProfile',
+  (roleName: string) => ({
+    InstanceProfileName: roleName,
+    RoleName: roleName,
+  }),
+);
+
+const deleteInstanceProfile = crudBuilder<IAM>(
+  'deleteInstanceProfile',
+  (roleName: string) => ({ InstanceProfileName: roleName, }),
+);
+
+const detachRoleToInstanceProfile = crudBuilder<IAM>(
+  'removeRoleFromInstanceProfile',
+  (roleName: string) => ({
+    InstanceProfileName: roleName,
+    RoleName: roleName,
+  }),
+);
+
+const getRole = crudBuilder<IAM>(
+  'getRole',
+  (roleName: string) => ({ RoleName: roleName, }),
+  (res: GetRoleCommandOutput) => res.Role,
+);
+
+const getAllRoles = paginateBuilder<IAM>(
+  paginateListRoles,
+  'Roles',
+);
+
+
+const updateRoleAssumePolicy = crudBuilder<IAM>(
+  'updateAssumeRolePolicy',
+  (roleName: string, assumeRolePolicyDocument: string) => ({
+    RoleName: roleName,
+    PolicyDocument: assumeRolePolicyDocument,
+  }),
+);
+
+const updateRoleDescription = crudBuilder<IAM>(
+  'updateRole',
+  (roleName: string, description?: string) => ({
+    RoleName: roleName,
+    Description: description
+  }),
+);
+
+const detachRolePolicy = crudBuilder<IAM>(
+  'detachRolePolicy',
+  (roleName: string, policyArn: string) => ({ RoleName: roleName, PolicyArn: policyArn, }),
+);
+const detachRolePolicies = (client: IAM, roleName: string, policyArns: string[]) => mapLin(
+  Promise.resolve(policyArns),
```

Comment:
Yes, since the `mapLin` receives is a `Promise<any []>` I'm passing the resolved promise directly. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve

Reply:
Oh, why not modify `mapLin` to also accept a bare array? eg, `arrProm: Promise<any[] | undefined> | any[]`

Reply:
Hmm well I thought it was easier this way, and no need to modify the function signature and add a new branch in the `mapLin` code for the `array` case or the `Promise` case.

Reply:
oh I see. ok if no change need it, it will be more natural the `mapLin` invocation